diff --git a/audio/AudioIn/UNIX/src/Android.mk b/audio/AudioIn/UNIX/src/Android.mk
index f88ef7f..e20e56c 100644
--- a/audio/AudioIn/UNIX/src/Android.mk
+++ b/audio/AudioIn/UNIX/src/Android.mk
@@ -25,7 +25,8 @@ LOCAL_STATIC_LIBRARIES := \
         libESR_Portable
 
 LOCAL_SHARED_LIBRARIES := \
-        liblog
+        liblog \
+	libmedia
 
 LOCAL_MODULE:= libSR_AudioIn
 
diff --git a/audio/AudioIn/UNIX/src/audioinwrapper.cpp b/audio/AudioIn/UNIX/src/audioinwrapper.cpp
index a30c017..079693d 100644
--- a/audio/AudioIn/UNIX/src/audioinwrapper.cpp
+++ b/audio/AudioIn/UNIX/src/audioinwrapper.cpp
@@ -95,16 +95,22 @@ int AudioOpen(void)
         audio_data = fopen ( file_name, "w" );
     #endif
 // TODO: get record buffer size from hardware.
+    android::AudioRecord::callback_t cbf = NULL;
     record = new android::AudioRecord(
                             AUDIO_SOURCE_DEFAULT,
                             sampleRate,
                             AUDIO_FORMAT_PCM_16_BIT,
                             (numChannels > 1) ? AUDIO_CHANNEL_IN_STEREO : AUDIO_CHANNEL_IN_MONO,
-                            8*1024);
-  
+                            8*1024,
+                            cbf,
+			    NULL,
+                            0,
+                            0);
+
+
   if (!record) return -1;
-  
-  return record->start() == NO_ERROR ? 0 : -1;
+
+  return record->start(android::AudioSystem::SYNC_EVENT_NONE,0) == NO_ERROR ? 0 : -1;
 #endif
 }
 
@@ -146,7 +152,7 @@ int AudioRead(short *buffer, int frame_count)
 #endif
 }
 
-int AudioSetVolume(int stream_type, int volume)
+int AudioSetVolume(audio_stream_type_t stream_type, float volume)
 {
 #if defined(USE_DEV_EAC_FILE)
   return 0;
@@ -155,7 +161,7 @@ int AudioSetVolume(int stream_type, int volume)
 #endif
 }
 
-int AudioGetVolume(int stream_type)
+int AudioGetVolume(audio_stream_type_t stream_type)
 {
 #if defined(USE_DEV_EAC_FILE)
   return 0;
@@ -203,12 +209,12 @@ int AudioSetStreamType(int stream_type)
   return -1;
 }
 
-int AudioSetVolume(int stream_type, int volume)
+int AudioSetVolume(audio_stream_type_t stream_type, float  volume)
 {
   return -1;
 }
 
-int AudioGetVolume(int stream_type)
+int AudioGetVolume(audio_stream_type_t stream_type)
 {
   return -1;
 }
